lib/spack/llnl/util/lang.py

```diff
@@ -374,6 +374,22 @@ def __iter__(self):
     return wrapper()
 
 
+def dedupe(sequence):
+    """Yields a stable de-duplication of an hashable sequence
+
+    Args:
+        sequence: hashable sequence to be de-duplicated
+
+    Returns:
+        stable de-duplication of the sequence
+    """
+    seen = set()
+    for x in sequence:
+        if x not in seen:
+            yield x
+            seen.add(x)
```

Comment:
@adamjstewart The keyword here is "stable" : `set` doesn't give you any guarantee on the order of the elements, while `dedupe` preserves the original order. In the context of library linking this is quite important to resolve symbols correctly.